libtenzir/builtins/operators/load_tcp.cpp

```diff
@@ -0,0 +1,961 @@
+  std::vector<std::thread> io_workers = {};
+  std::shared_ptr<boost::asio::io_context> io_ctx = {};
+  std::optional<boost::asio::ip::tcp::socket> socket = {};
+
+  // Everything required for listening for connections.
+  std::optional<boost::asio::ip::tcp::acceptor> acceptor = {};
+
+  // Everything needed for back pressure handling.
+  static constexpr auto max_buffered_batches = size_t{20};
+  std::queue<Elements> buffer = {};
+  caf::typed_response_promise<Elements> read_rp = {};
+  std::queue<caf::typed_response_promise<void>> write_rps = {};
+
+  // State we need to keep for each peer.
+  struct connection_state : std::enable_shared_from_this<connection_state> {
+    connection_state() = default;
+    connection_state(const connection_state&) = delete;
+    connection_state(connection_state&&) = delete;
+    auto operator=(const connection_state&) -> connection_state& = delete;
+    auto operator=(connection_state&&) -> connection_state& = delete;
+
+    ~connection_state() noexcept {
+      next_emit_metrics.dispose();
+      emit_metrics(nullptr);
+    }
+
+    static constexpr auto max_queued_chunks = size_t{10};
+    static constexpr auto read_buffer_size = size_t{65'536};
+
+    std::optional<boost::asio::ip::tcp::socket> socket = {};
+    std::optional<boost::asio::ssl::context> ssl_ctx = {};
+    std::optional<boost::asio::ssl::stream<boost::asio::ip::tcp::socket&>>
+      tls_socket = {};
+    pipeline_executor_actor pipeline_executor = {};
+
+    // The mutex is protecting the queue of chunks and the response promise, as
+    // they're both used from the Asio-managed thread pool.
+    mutable std::mutex mutex = {};
+    std::queue<chunk_ptr> chunks = {};
+    caf::typed_response_promise<chunk_ptr> rp = {};
+
+    metrics_receiver_actor metrics_receiver = {};
+    uint64_t operator_id = {};
+    uint64_t reads = {};
+    uint64_t bytes_read = {};
+    caf::disposable next_emit_metrics = {};
+
+    auto emit_metrics(connection_manager_actor<Elements>::pointer self)
+      -> void {
+      auto metric = record{
+        {"handle", fmt::to_string(socket->native_handle())},
+        {"reads", std::exchange(reads, {})},
+        {"writes", uint64_t{0}},
+        {"bytes_read", std::exchange(bytes_read, {})},
+        {"bytes_written", uint64_t{0}},
+      };
+      caf::anon_send(metrics_receiver, operator_id, tcp_metrics_id,
+                     std::move(metric));
+      if (self) {
+        next_emit_metrics = detail::weak_run_delayed(
+          self, defaults::metrics_interval,
+          [self, connection = this->shared_from_this()] {
+            connection->emit_metrics(self);
+          });
+      }
+    }
+
+    auto async_read(connection_manager_actor<Elements>::pointer self,
+                    shared_diagnostic_handler diagnostics) -> void {
+      // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays)
+      auto read_buffer
+        = std::make_unique<chunk::value_type[]>(read_buffer_size);
```

Comment:
I wasn't sure about stability on `.data()` on vector when implementing this. This could be changed, but there's little reason to make that change now.